ddl/index.go

```diff
@@ -324,184 +326,275 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
 	return errors.Trace(err)
 }
 
-func (d *ddl) fetchRowColVals(txn kv.Transaction, t table.Table, batchOpInfo *indexBatchOpInfo, seekHandle int64) error {
-	cols := t.Cols()
-	idxInfo := batchOpInfo.tblIndex.Meta()
-
-	err := d.iterateSnapshotRows(t, txn.StartTS(), seekHandle,
+func (d *ddl) fetchRowColVals(txn kv.Transaction, t table.Table, taskOpInfo *indexTaskOpInfo, handleInfo *handleInfo) (
+	[]*indexRecord, *taskResult) {
+	handleCnt := defaultTaskHandleCnt
+	rawRecords := make([][]byte, 0, handleCnt)
+	idxRecords := make([]*indexRecord, 0, handleCnt)
+	ret := &taskResult{doneHandle: handleInfo.startHandle}
+	err := d.iterateSnapshotRows(t, txn.StartTS(), handleInfo.startHandle,
 		func(h int64, rowKey kv.Key, rawRecord []byte) (bool, error) {
-			rowMap, err := tablecodec.DecodeRow(rawRecord, batchOpInfo.colMap)
-			if err != nil {
-				return false, errors.Trace(err)
-			}
-			idxVal := make([]types.Datum, 0, len(idxInfo.Columns))
-			for _, v := range idxInfo.Columns {
-				col := cols[v.Offset]
-				idxVal = append(idxVal, rowMap[col.ID])
-			}
-
-			indexRecord := &indexRecord{handle: h, key: rowKey, vals: idxVal}
-			batchOpInfo.idxRecords = append(batchOpInfo.idxRecords, indexRecord)
-			if len(batchOpInfo.idxRecords) == defaultSmallBatchCnt {
+			rawRecords = append(rawRecords, rawRecord)
+			indexRecord := &indexRecord{handle: h, key: rowKey}
+			idxRecords = append(idxRecords, indexRecord)
+			if len(idxRecords) == handleCnt || handleInfo.isFinished(h) {
 				return false, nil
 			}
 			return true, nil
 		})
 	if err != nil {
-		return errors.Trace(err)
-	} else if len(batchOpInfo.idxRecords) == 0 {
-		return nil
+		ret.err = errors.Trace(err)
+		return nil, ret
 	}
 
-	count := len(batchOpInfo.idxRecords)
-	batchOpInfo.addedCount += int64(count)
-	batchOpInfo.handle = batchOpInfo.idxRecords[count-1].handle
-	return nil
+	ret.count = len(idxRecords)
+	if ret.count > 0 {
+		ret.doneHandle = idxRecords[ret.count-1].handle
+	}
+	// Be sure to do this operation only once.
+	if !handleInfo.isSent {
+		// Notice to start the next task operation.
+		taskOpInfo.nextCh <- ret.doneHandle
+		// Record the last handle.
+		// Ensure that the handle scope of the task doesn't change,
+		// even if the transaction retries it can't effect the other tasks.
+		handleInfo.endHandle = ret.doneHandle
+		handleInfo.isSent = true
+	}
+	if ret.count == 0 {
```

Comment:
Does `ret.count` always equal to `len(idxRecords)`? If so, you needn't record it.